core/src/test/java/com/graphhopper/routing/weighting/custom/SpeedCalculatorTest.java

```diff
@@ -0,0 +1,180 @@
+/*
+ *  Licensed to GraphHopper GmbH under one or more contributor
+ *  license agreements. See the NOTICE file distributed with this work for
+ *  additional information regarding copyright ownership.
+ *
+ *  GraphHopper GmbH licenses this file to you under the Apache License,
+ *  Version 2.0 (the "License"); you may not use this file except in
+ *  compliance with the License. You may obtain a copy of the License at
+ *
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+package com.graphhopper.routing.weighting.custom;
+
+import com.graphhopper.routing.profiles.EnumEncodedValue;
+import com.graphhopper.routing.profiles.RoadClass;
+import com.graphhopper.routing.profiles.RoadEnvironment;
+import com.graphhopper.routing.util.CustomModel;
+import com.graphhopper.routing.util.EncodingManager;
+import com.graphhopper.routing.util.FlagEncoder;
+import com.graphhopper.storage.GraphBuilder;
+import com.graphhopper.util.EdgeIteratorState;
+import org.junit.jupiter.api.Test;
+
+import java.util.HashMap;
+import java.util.Map;
+
+import static org.junit.jupiter.api.Assertions.*;
+
+class SpeedCalculatorTest {
+
+    private final EncodingManager em;
+    private final EdgeIteratorState edge;
+
+    SpeedCalculatorTest() {
+        // in this test we use the same edge for every test and check the calculated speed for different custom models
+        // and depending on additional properties of the edge we set in the tests
+        em = EncodingManager.create("car");
+        edge = new GraphBuilder(em).create().edge(0, 1, 1000, true);
+    }
+
+    @Test
+    public void maxSpeedFallback() {
+        // default speed for car is 60km/h
+        assertEquals(60, calcSpeed(edge, new CustomModel()));
+        // use global fallback to limit speed
+        assertEquals(30, calcSpeed(edge, new CustomModel().setMaxSpeedFallback(30.0)));
+    }
+
+    @Test
+    public void maxSpeed() {
+        // here we use max_speed to limit speed
+        CustomModel model = new CustomModel();
+        Map<String, Object> map = new HashMap<>();
+        map.put("*", 25);
+        model.getMaxSpeed().put(RoadClass.KEY, map);
+        assertEquals(25, calcSpeed(edge, model));
+        // ... but setting it to a higher value than the default has no effect
+        map.put("*", 70);
+        assertEquals(60, calcSpeed(edge, model));
+        // also our edge might or might not be affected
+        // todonow: why has encoding manager a road class encoded value by default?
```

Comment:
Might be a bit unrelated here, but is it a good idea to hard-code some default encoded values?

Reply:
I think we need default EncodedValues yes, see #1805 where I introduced them. Otherwise instructions and similar things might be too complicated or even impossible to implement. Or did you mean something else?

Reply:
Ah ok, yes I thought maybe these defaults can be moved to config-example.yml or something. But I was just wondering.

Reply:
This is not possible unfortunately as then the pure Java version would throw errors.
Also assume you would need to add road_class, roundabout, road_environment etc everytime, before using GraphHopper.

Reply:
> This is not possible unfortunately as then the pure Java version would throw errors.

Ok the equivalent would be setting defaults in `GraphHopperConfig` for example.

> Also assume you would need to add road_class, roundabout, road_environment etc everytime, before using GraphHopper.

Hm, I think this would be easier to understand (GraphHopper has the encoded values I am telling it to have and I do not have to check whether there is some hard-coded stuff), but yes maybe a bit harder to use. Anyway lets keep it as it is I was just thinking out loud.

Reply:
> Anyway lets keep it as it is I was just thinking out loud.

I'm still trying to understand this :) as I also did not find it very clean. When you configure this e.g. in GraphHopperConfig and you are allowed to remove `road_class` how would you prevent the instruction code where `road_class` is required from being called (and broken)?

Reply:
Ok if the instructions code has to use `road_class` then it somehow has to be the default. Can the instruction code skip it in case it is not there? But maybe some encoded values are simply always required and different than others. Also so far we are not using `GraphHopperConfig` to specify all the default values, but they are still somewhere in `GraphHopper` etc. I think in the mid/long run we should aim for GH (the library) not having any hard-coded stuff (just the mechanim), i.e. separate the code from the configuration as much as possible. But maybe this is something for/after #1895?

Reply:
> Can the instruction code skip it in case it is not there?

Yes but this is already rather ugly (see get_off_bike if-else stuff), especially when we improve instructions we will need not only need road_class. Likely the instruction creation is something that needs to be more pluggable.

> But maybe some encoded values are simply always required and different than others

Yes, definitely. Like roundabout, road_environment or road_class(_link) and maybe later `lanes` (but only for car).

> I think in the mid/long run we should aim for GH (the library) not having any hard-coded stuff (just the mechanism), i.e. separate the code from the configuration as much as possible. 

Yes, sure. When I think about it it even gets more complicated (for non-road use cases) and having these default values defined (in EncodingManager) is then definitely not desired.

Probably half of the issue is resolved when we remove the FlagEncoder because then we define collections of encoded values and automatically add lanes&road_class for the car collection or whatever.